backend/triplestore/rdf_field.py

```diff
@@ -0,0 +1,246 @@
+        utils.replace_triples(
+            g,
+            self._stored_triples(instance),
+            self._triples_to_store(instance, value)
+        )
+
+    def clear(self, instance) -> None:
+        '''
+        Clear the field's data in a graph for a given instance of the model.
+        
+        This is used as cleanup when deleting the model instance.
+
+        Parameters:
+            instance: the model instance
+        '''
+        g = self.get_graph(instance)
+        for triple in self._stored_triples(instance):
+            g.remove(triple)
+
+    def _stored_triples(self, instance) -> utils.Triples:
+        '''
+        Extract the set of relevant triples from a graph representation
+        '''
+        raise NotImplementedError()
+
+    def _triples_to_store(self, instance, value) -> utils.Triples:
+        '''
+        Triples that should be stored to represent the modelled value
+        '''
+        raise NotImplementedError()
+
+
+class RDFPredicateField(RDFField):
+    '''
+    Models a predicate, i.e. a property + object in the graph. This translates to a
+    boolean value in the model.
+
+    Note that unlike the RDFPropertyField (and derivatives), this field does not model all
+    triples with the provided property. That is, if `A` is an instance of a model with
+    `RDFPredicateField(B, C)`, the value of the field represents whether the triple
+    `(A, B, C)` is present in a graph. The presence of another triple `(A, B, D)` is
+    considered irrelevant, and will not be affected when you save the field.
+    '''
+
+    def __init__(self, property: URIRef, object: Identifier, **kwargs):
+        super().__init__(**kwargs)
+        self.property = property
+        self.object = object
+
+    def get(self, instance):
+        return any(self._stored_triples(instance))
+
+    def _stored_triples(self, instance):
+        g = self.get_graph(instance)
+        subject = instance.uri
+        return g.triples((subject, self.property, self.object))
+    
+    def _triples_to_store(self, instance, value):
+        if value:
+            subject = instance.uri
+            return [(subject, self.property, self.object)]
+        else:
+            return []
+
+class RDFPropertyField(RDFField):
+    '''
+    Models a property relationship. For a given subject, this field represents all triples
+    (s, p, o), where s is the node represented by the model instance, and p is the field's
+    configured property.
+    '''
+
+    def __init__(self, property: URIRef, **kwargs):
+        super().__init__(**kwargs)
+        self.property = property
+
+    def get(self, instance):
+        return [
+            self.node_to_value(o)
+            for (s, p, o) in self._stored_triples(instance)
+        ]
+
+    def node_to_value(self, node: Identifier):
+        '''
+        Convert an object node to its modelled value
+        '''
+        if isinstance(node, Literal):
+            return node.value
```

Comment:
I understand how this works if `node` is a `Literal`, but what about other cases? Are these supported at all?

Reply:
I've added some clarification to the docstring. If the node is a `BNode` or `URIRef`, it will be returned as-is. Alternatively, you could create a subclass that handles values some other way.